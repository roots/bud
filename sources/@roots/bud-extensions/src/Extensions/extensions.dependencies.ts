import {bind, lodash} from '@roots/bud-support'

export {bind}
export const {
  isEqual,
  isString,
  get,
  isFunction,
  isNull,
  isUndefined,
} = lodash
