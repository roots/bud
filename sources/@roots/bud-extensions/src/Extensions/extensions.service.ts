import {
  Extension,
  Extensions as Base,
  Framework,
  Service,
} from '@roots/bud-framework'

import {Controller} from '../Controller'
import {bind} from './extensions.dependencies'

/**
 * Extensions Service
 *
 * @remarks
 * Manages extension controllers
 *
 * @public
 */
export class Extensions extends Service implements Base {
  /**
   * Extensions queued for registration
   *
   * @public
   */
  public queue = []

  /**
   * Controller factory
   *
   * @public
   */
  public makeController(
    extension: Extension.Module | Promise<Extension.Module>,
  ): Controller {
    const controller = new Controller(this.app, extension)
    return controller
  }

  @bind
  public async setController(extension: Extension.Module): Promise<void> {
    const controller = this.makeController(extension)
    this.set(controller.name, controller)
  }

  /**
   * @override
   * @public
   */
  @bind
  public async booted(): Promise<void> {
    /**
     * Handle in-built extensions
     */
    await Promise.all(
      this.getEntries().map(async ([key, extension]) => {
        this.setController(extension)
        this.log('success', {message: `${key} instantiated`})
      }),
    )

    await this.registerExtensions()
    await this.bootExtensions()
    await this.injectExtensions()
  }

  /**
   * Inject extension modules
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public async injectExtensions() {
    if (this.app.store.is('features.inject', false)) {
      this.log('log', 'injection disabled')
      return
    }

    try {
      const modules = Object.values(this.app.project.peers.modules)
        .filter(Boolean)
        .filter(({bud}) => bud?.type === 'extension')

      await Promise.all(
        modules.map(async record => {
          await this.importExtension(record)
        }),
      )

      await modules.reduce(async (promised: Promise<void>, record) => {
        await promised
        await this.registerExtension(this.get(record.name))
      }, Promise.resolve())

      await modules.reduce(async (promised, record) => {
        await promised
        await this.bootExtension(this.get(record.name))
      }, Promise.resolve())
    } catch (e) {
      this.app.error(e)
    }
  }

  @bind
  public async importExtension(
    extension: Record<string, any>,
  ): Promise<void> {
    this.log('log', `importing ${extension.name}`)
    const importedModule = await import(extension.name)
    const importedExtension: Extension.Module = importedModule.default
      ? importedModule.default
      : importedModule

    if (this.has(importedExtension.name)) return
    await this.setController(importedExtension)
  }

  /**
   * @public
   */
  @bind
  public async registerExtension(extension: Controller): Promise<void> {
    try {
      if (!extension) return
      this.app.log('registering', extension.name)

      await extension.mixin()
      await extension.api()
      await extension.register()
    } catch (err) {
      this.app.log(extension)
      throw new Error(err)
    }
  }

  /**
   * @public
   */
  @bind
  public async bootExtension(extension: Controller): Promise<void> {
    try {
      if (!extension) return
      this.app.log('booting', extension.name)

      await extension.boot()
    } catch (err) {
      throw new Error(err)
    }
  }

  /**
   * @public
   */
  @bind
  public async registerExtensions(): Promise<void> {
    this.log('time', 'registering')

    await this.getEntries().reduce(
      async (promised, [_key, controller]) => {
        await promised
        await this.registerExtension(controller)
      },
      Promise.resolve(),
    )

    this.log('timeEnd', 'registering')
  }

  /**
   * @public
   */
  @bind
  public async bootExtensions(): Promise<void> {
    this.log('time', 'booting')
    await this.getEntries().reduce(async (promised, [key, controller]) => {
      await promised
      await this.bootExtension(controller)
    }, Promise.resolve())

    this.log('timeEnd', 'booting')
  }

  /**
   * Add a {@link Controller} to the container
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public async add(extension: Extension.Module): Promise<void> {
    if (this.has(extension.name)) {
      this.log('info', `${extension.name} already exists. skipping.`)
      return
    }

    await this.setController(extension)
    await this.registerExtension(this.get(extension.name))
    await this.bootExtension(this.get(extension.name))
  }

  /**
   * Queue an extension to be added to the container before the build process.
   *
   * @remarks
   * Useful for extensions which cannot be added in an awaitable context (like a user config)
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public enqueue(extension: Extension.Module): Framework {
    this.queue.push(extension)
    return this.app
  }

  /**
   * @public
   */
  @bind
  public async processQueue(): Promise<void> {
    if (!this.queue.length) return
    await Promise.all(this.queue.map(this.add))
    this.queue = []
  }

  /**
   * Returns an array of plugin instances which have been registered to the
   * container and are set to be used in the compilation
   *
   * @returns An array of plugin instances
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public async make(): Promise<
    {
      [key: string]: any
      apply: CallableFunction
    }[]
  > {
    this.log('time', 'extensions.make')

    await this.processQueue()

    const plugins = this.getValues()
      .filter(controller => controller._module.make)
      .map((controller: Controller) => {
        const result = controller.make()

        if (!result) {
          this.log(
            'log',
            `${controller.name} will not be used in the compilation`,
          )

          return result
        }

        this.log(
          'success',
          `${controller.name} will be used in the compilation`,
        )

        return result
      })
      .filter(Boolean)

    this.log('timeEnd', 'extensions.make')

    return plugins
  }
}
