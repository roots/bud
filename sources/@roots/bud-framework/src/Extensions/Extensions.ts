import {Framework} from '../Framework'
import {Service} from '../Service'
import {CompilerPlugin, Controller, Module} from './Extension'

/**
 * Extensions Service
 *
 *  @public
 */
export interface Extensions extends Service {
  /**
   * Extensions to be processed before build
   *
   * @public
   */
  queue: Array<CompilerPlugin | Module>

  /**
   * Add a {@link Controller} to the container
   *
   * @public
   */
  add(extension: CompilerPlugin | Module): Promise<void>

  /**
   * Queue an extension to be added to the container before the build process.
   *
   * @remarks
   * Useful for extensions which cannot be added in an awaitable context (like a user config)
   *
   * @public
   * @decorator `@bind`
   */
  enqueue(extension: CompilerPlugin | Module): Framework

  /**
   * Auto install and register discovered extensions
   * @public
   */
  injectExtensions(): Promise<void>

  /**
   * Call {@link Controller.register} on all {@link Controller}s
   *
   * @public
   */
  registerExtensions(): Promise<void>

  /**
   * Call an {@link Controller.register}
   *
   * @public
   */
  registerExtension(extension: Controller): Promise<void>

  /**
   * Boot event for all extensions
   *
   * @public
   */
  bootExtensions(): Promise<void>

  /**
   * Returns an array of plugin instances which have been registered to the
   * container and are set to be used in the compilation
   *
   * @returns An array of plugin instances
   *
   * @public
   */
  make(): Promise<{[key: string]: any; apply: CallableFunction}[]>

  /**
   * @public
   */
  processQueue(): Promise<void>
}
