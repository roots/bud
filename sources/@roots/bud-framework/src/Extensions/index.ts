export {Module} from './Extension'
export {Extensions} from './Extensions'
export {Controller} from './Extension/controller.interface'
