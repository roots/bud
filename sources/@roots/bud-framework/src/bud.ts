import {bind} from '@roots/bud-support/decorators'
import isFunction from '@roots/bud-support/lodash/isFunction'
import isNull from '@roots/bud-support/lodash/isNull'
import isString from '@roots/bud-support/lodash/isString'
import isUndefined from '@roots/bud-support/lodash/isUndefined'

import {bootstrap, LIFECYCLE_EVENT_MAP} from './lifecycle/bootstrap.js'
import type * as methods from './methods/index.js'
import type {Module} from './module.js'
import type {Service} from './service.js'
import type ConsoleBuffer from './services/console.js'
import type FS from './services/fs.js'
import type * as Options from './types/options/index.js'
import type * as Registry from './types/registry/index.js'
import type Hooks from './types/services/hooks/index.js'
import type * as Services from './types/services/index.js'
import Value from './value.js'

/**
 * Framework abstract
 */
export class Bud {
  /**
   * Context
   *
   * @public
   */
  public context:
    | Options.CommandContext
    | Options.CLIContext
    | Options.Context

  public isCLI(): this is Bud & {context: Options.CommandContext} {
    return `args` in this.context
  }

  /**
   * Implementation
   *
   * @public
   */
  public implementation: new () => Bud

  /**
   * Compilation mode
   *
   * @remarks
   * Either `production` or `development`.
   *
   * @readonly
   * @defaultValue `production`
   * @public
   */
  public get mode(): `development` | `production` {
    return this.context.mode ?? `production`
  }

  /**
   * Name
   *
   * @readonly
   * @public
   */
  public get label() {
    return this.context.label
  }

  /**
   * Parent {@link Bud} instance
   *
   * @readonly
   * @public
   */
  public get root() {
    return this.context.root ?? this
  }

  /**
   * True when {@link Bud.mode} is `production`
   *
   * @public
   */
  public get isProduction(): boolean {
    return this.mode == `production`
  }

  /**
   * True when {@link Bud.mode} is `development`
   *
   * @public
   */
  public get isDevelopment(): boolean {
    return this.mode == `development`
  }

  /**
   * True when current instance is the parent instance
   *
   * @readonly
   * @public
   */
  public get isRoot(): boolean {
    return this.root.label === this.label
  }

  /**
   * True when current instance is a child instance
   *
   * @readonly
   * @public
   */
  public get isChild(): boolean {
    return this.root.label !== this.label
  }

  /**
   * {@link Bud} instances
   *
   * @public
   */
  public children?: Record<string, Bud>

  /**
   * True when child compilers
   *
   * @readonly
   * @public
   */
  public get hasChildren(): boolean {
    return (
      !isUndefined(this.children) &&
      !isNull(this.children) &&
      Object.entries(this.children).length > 0
    )
  }

  public consoleBuffer: ConsoleBuffer

  public fs: FS

  public module: Module

  public services: Array<string> = []

  public api: Services.Api

  public build: Services.Build.Service

  public cache: Services.Cache.Service

  public compiler: Services.Compiler.Service

  public dashboard: Services.Dashboard.Service

  public env: Services.Env

  public extensions: Services.Extensions.Service

  public hooks: Hooks

  public project: Services.Project.Service

  public server: Services.Server.Service

  public declare after: methods.after

  public declare maybeCall: methods.maybeCall

  public declare close: methods.close

  public declare container: methods.container

  public declare get: methods.get

  public declare glob: methods.glob

  public declare globSync: methods.globSync

  public declare path: methods.path

  public declare pipe: methods.pipe

  public declare processConfigs: methods.processConfigs

  public declare publicPath: methods.publicPath

  public declare relPath: methods.relPath

  public declare run: methods.run

  public declare setPath: methods.setPath

  public declare setPublicPath: methods.setPublicPath

  public declare sequence: methods.sequence

  public declare sequenceSync: methods.sequenceSync

  public declare sh: methods.sh

  public declare tap: methods.tap

  public declare tapAsync: methods.tapAsync

  public declare when: methods.when

  public declare bindMethod: methods.bindMethod

  /**
   * @deprecated - use {@link bud.fs.json | bud.fs.json}
   */
  public json: FS['json']

  /**
   * @deprecated - use {@link bud.fs.yml | bud.fs.yml}
   */
  public yml: FS['yml']

  /**
   * Value helper
   *
   * @public
   */
  public value = Value

  /**
   * Creates a child with `bud.create` but returns the parent instance
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public async make(
    request: Partial<Options.Context> | string,
    tap?: (app: Bud) => Promise<unknown>,
  ) {
    if (!this.isRoot)
      return this.fatal(
        new Error(
          `Child instances should be produced from the root context`,
        ),
      )

    const context: Options.Context = isString(request)
      ? {...this.context, label: request, root: this}
      : {...this.context, ...request, root: this}

    if (
      this.isCLI() &&
      !isUndefined(this.context.args.filter) &&
      !this.context.args.filter.includes(context.label)
    ) {
      this.log(
        `skipping child instance based on --filter flag:`,
        context.label,
      )
      return this
    }

    if (this.children && this.children[context.label]) {
      this.log(`returning requested child instance:`, context.label)
      return this.get(context.label)
    }

    this.log(`instantiating new bud instance`)
    const child = await new Bud().lifecycle(context)

    if (!this.children) this.children = {[context.label]: child}
    else this.children[context.label] = child

    this.get(context.label).hooks.on(
      `build.dependencies`,
      typeof request !== `string` && request.dependsOn
        ? request.dependsOn
        : Object.values(this.children)
            .map(({label}) => label)
            .filter(label => label !== context.label),
    )

    if (tap) await tap(this.get(context.label))
    await this.get(context.label)?.api.processQueue()

    return this
  }

  @bind
  public async lifecycle(context: Options.Context): Promise<Bud> {
    await bootstrap.bind(this)({...context})

    Object.entries(LIFECYCLE_EVENT_MAP).map(
      ([eventHandle, callbackName]: [
        keyof Registry.EventsStore,
        keyof Service,
      ]) =>
        this.services
          .map(service => [service, this[service]])
          .map(([label, service]) => {
            if (!service) {
              this.error(`service not found: ${label}`, this.services)
            }

            if (!isFunction(service[callbackName])) return

            this.hooks.action(
              eventHandle,
              service[callbackName].bind(service),
            )
            this.success(
              `registered service callback:`,
              `${label}.${callbackName}`,
            )
          }),
    )

    await [
      `init`,
      `bootstrap`,
      `bootstrapped`,
      `register`,
      `registered`,
      `boot`,
      `booted`,
    ].reduce(async (promised, event: keyof Registry.EventsStore) => {
      await promised
      await this.hooks.fire(event)
    }, Promise.resolve())

    return this
  }

  /**
   * Log a message
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public log(...messages: any[]) {
    this.context.logger.log(...messages)

    return this
  }

  /**
   * Log an `info` level message
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public info(...messages: any[]) {
    this.context.logger.info(...messages)

    return this
  }

  /**
   * Log a `success` level message
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public success(...messages: any[]) {
    this.context.logger.success(...messages)
    return this
  }

  /**
   * Log a `warning` level message
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public warn(...messages: any[]) {
    this.context.logger.warn(...messages)

    return this
  }

  /**
   * Log an error.
   *
   * @remarks
   * In `production` this error is treated as fatal
   * and will kill the process.
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public error(...messages: Array<any>): Bud {
    this.context.logger.error(...messages)
    return this
  }

  /**
   * Log and display an error.
   *
   * @remarks
   * This will always kill the process
   *
   * @public
   * @decorator `@bind`
   * @throws fatal error
   */
  @bind
  public fatal(error: Error) {
    this.context.logger.fatal(error)
    return this
  }
}
