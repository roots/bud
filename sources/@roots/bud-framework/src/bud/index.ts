import type {
  Api,
  Build,
  Cache,
  Compiler,
  Context,
  Dashboard,
  Env,
  Extensions,
  Hooks,
  Project,
  Server,
  Service,
} from '@roots/bud-framework'

import methods from '@roots/bud-framework/methods'
import {bind} from '@roots/bud-support/decorators/bind'
import {BudError, InputError} from '@roots/bud-support/errors'
import isFunction from '@roots/bud-support/isFunction'
import isString from '@roots/bud-support/isString'
import isUndefined from '@roots/bud-support/isUndefined'
import logger from '@roots/bud-support/logger'

import type {FS} from '../fs.js'
import type {Module} from '../module.js'
import type {Notifier} from '../notifier.js'
import type {EventsStore} from '../registry/index.js'

import {bootstrap} from '../bootstrap.js'

/**
 * Bud core class
 */
export class Bud {
  /**
   * Promised tasks
   */
  public promised: Array<(bud: Bud) => Promise<any>> = []

  public declare addConfig: typeof methods.addConfig

  public declare after: typeof methods.after

  public declare api: Service & Api

  public declare bindFacade: typeof methods.bindFacade

  public declare build: Service & Build

  public declare cache: Service & Cache

  /**
   * {@link Bud} instances
   */
  public declare children: Record<string, Bud>

  public declare close: typeof methods.close

  public declare compiler?: Service & Compiler

  public declare container: typeof methods.container

  public declare context: Context

  public declare dashboard: Service & Dashboard

  public declare env: Service & Env

  public declare extensions: Service & Extensions

  public declare fs: FS

  public declare get: typeof methods.get

  public declare glob: typeof methods.glob

  public declare globSync: typeof methods.globSync

  public declare hooks: Service & Hooks

  /**
   * {@link Bud} Implementation
   * @internal
   */
  public declare implementation: new () => Bud

  /**
   * @deprecated Use {@link bud.fs}
   * @readonly
   */
  public declare json: FS['json']

  public declare maybeCall: typeof methods.maybeCall

  public declare module: Module

  public declare notifier: Notifier

  public declare path: typeof methods.path

  public declare pipe: typeof methods.pipe

  public declare processConfigs: typeof methods.processConfigs

  public declare project: Project

  public declare publicPath: typeof methods.publicPath

  public declare relPath: typeof methods.relPath

  public declare run: typeof methods.run

  public declare sequence: typeof methods.sequence

  public declare sequenceSync: typeof methods.sequenceSync

  public declare server?: Service & Server

  public declare setPath: typeof methods.setPath

  public declare setPublicPath: typeof methods.setPublicPath

  public declare sh: typeof methods.sh

  public declare tap: typeof methods.tap

  public declare tapAsync: typeof methods.tapAsync

  public declare when: typeof methods.when

  /**
   * @deprecated Use {@link Bud.fs}
   * @readonly
   */
  public declare yml: FS['yml']

  /**
   * True when child compilers
   * @readonly
   */
  public get hasChildren(): boolean {
    return this.children && Object.entries(this.children).length > 0
  }

  /**
   * True when current instance is a child instance
   * @readonly
   */
  public get isChild(): boolean {
    return this.root?.context?.label !== this.context?.label
  }

  /**
   * True when {@link Bud.mode} is `development`
   * @readonly
   */
  public get isDevelopment(): boolean {
    return this.mode === `development`
  }

  /**
   * True when {@link Bud.mode} is `production`
   * @readonly
   */
  public get isProduction(): boolean {
    return this.mode === `production`
  }

  /**
   * True when current instance is the parent instance
   * @readonly
   */
  public get isRoot(): boolean {
    return this.root?.context?.label === this.context?.label
  }

  /**
   * Label
   * @readonly
   */
  public get label() {
    return this.context?.label ?? `bud`
  }

  /**
   * Compilation mode
   *
   * @remarks
   * Either `production` or `development`.
   *
   * @readonly
   * @defaultValue `production`
   */
  public get mode(): `development` | `production` {
    return this.context?.mode ?? `production`
  }

  /**
   * Constructor
   */
  public constructor(context?: Context) {
    if (context) this.context = {...context}

    this.set(`implementation`, this.constructor as any)

    Object.entries(methods).map(([k, v]) =>
      this.set(k as any, v.bind(this)),
    )
  }

  /**
   * Boot application services
   */
  @bind
  public async boot() {
    await this.executeServiceCallbacks(`boot`).catch(this.catch)
  }

  /**
   * Bootstrap the application
   */
  @bind
  public async bootstrap() {
    await this.executeServiceCallbacks(`bootstrap`).catch(this.catch)
  }

  /**
   * Error handler
   */
  @bind
  public catch(error: Error): never {
    const normalizedError = BudError.normalize(error)
    if (!normalizedError.instance && this?.isChild)
      normalizedError.instance = this.label
    throw normalizedError
  }

  /**
   * Execute service callbacks for a given stage
   * @internal
   */
  @bind
  public async executeServiceCallbacks(
    stage: `${keyof EventsStore & string}`,
  ): Promise<void> {
    await this.resolvePromises().catch(this.catch)
    await this.hooks.fire(stage, this)
  }

  /**
   * Bud initialize
   */
  @bind
  public async initialize(context?: Context): Promise<Bud> {
    if (context) this.context = {...(this.context ?? {}), ...context}

    await bootstrap(this).catch(this.catch)

    await this.bootstrap().catch(this.catch)
    await this.register().catch(this.catch)
    await this.boot().catch(this.catch)
    await this.executeServiceCallbacks(`config.before`).catch(this.catch)

    return this
  }

  /**
   * Creates a child and returns the parent instance
   */
  @bind
  public async make(
    request: Partial<Context> | string,
    setupFn?: (app: Bud) => Promise<unknown>,
  ) {
    if (!this.isRoot) {
      return this.catch(
        InputError.normalize(
          `bud.make: must be called from the root context`,
        ),
      )
    }

    const context: Context = isString(request)
      ? {...this.context, label: request, root: this}
      : {...this.context, ...request, root: this}

    if (isUndefined(context.label)) {
      return this.catch(
        InputError.normalize(`bud.make: context.label must be a string`),
      )
    }

    if (
      !isUndefined(this.context.filter) &&
      !this.context.filter.includes(context.label)
    ) {
      logger
        .scope(`make`)
        .log(
          `skipping child instance based on --filter flag:`,
          context.label,
        )

      return this
    }

    if (!this.children) this.children = {}

    if (this.children[context.label]) {
      return this.catch(
        InputError.normalize(
          `bud.make: child instance ${context.label} already exists`,
        ),
      )
    }

    logger.scope(`make`).log(`instantiating ${context.label}`)

    this.children[context.label] =
      await new this.implementation().initialize({
        ...context,
      })
    if (setupFn) await setupFn(this.children[context.label])

    await this.children[context.label].resolvePromises().catch(this.catch)

    if (typeof request !== `string` && request.dependsOn) {
      this.get(context.label)?.hooks.on(
        `build.dependencies`,
        request.dependsOn,
      )
    }

    return this
  }

  /**
   * Await all promised tasks
   */
  @bind
  public promise(promise: (bud: Bud) => Promise<unknown>): Bud {
    this.promised.push(promise)
    return this
  }

  @bind
  public async resolvePromises() {
    if (this.promised.length === 0) return

    const promises = this.promised

    this.promised = []

    await promises.reduce(async (promised, promise) => {
      await promised
      await promise(this)
    }, Promise.resolve())
  }

  /**
   * Register application services
   */
  @bind
  public async register() {
    await this.executeServiceCallbacks(`register`).catch(this.catch)
  }

  /**
   * Parent {@link Bud} instance
   * @readonly
   */
  public get root(): Bud {
    return this.context?.root ?? this
  }

  /**
   * Set a value on the current instance
   * @param key - key
   * @param value - value
   * @param bind - bind value to current instance (default: true, if bindable)
   */
  @bind
  public set<K extends `${keyof Bud & string}`>(
    key: K,
    input: Bud[K],
    bind: boolean = true,
  ): Bud {
    const toBind = bind && isFunction(input) && `bind` in input

    const value = toBind ? input.bind(this) : input

    Object.defineProperty(this, key, {
      configurable: true,
      enumerable: true,
      value,
      writable: true,
    })

    return this
  }

  /**
   * Log message
   * @deprecated Import logger instance from `@roots/bud-support/logger`
   */
  @bind
  public log(...messages: Array<any>) {
    logger.scope(this.label).log(...messages)
    return this
  }

  /**
   * Log success
   * @deprecated Import logger instance from `@roots/bud-support/logger`
   */
  @bind
  public success(...messages: Array<any>) {
    logger.scope(this.label).log(...messages)
    return this
  }
  /**
   * Log warning
   * @deprecated Import logger instance from `@roots/bud-support/logger`
   */
  @bind
  public warn(...messages: Array<any>) {
    logger.scope(this.label).warn(...messages)
    return this
  }

  /**
   * Log info
   * @deprecated Import logger instance from `@roots/bud-support/logger`
   */
  @bind
  public info(...messages: Array<any>) {
    logger.scope(this.label).info(...messages)
    return this
  }

  /**
   * Log error
   * @deprecated Import logger instance from `@roots/bud-support/logger`
   */
  @bind
  public error(...messages: Array<any>) {
    logger.scope(this.label).error(...messages)
    return this
  }
}
