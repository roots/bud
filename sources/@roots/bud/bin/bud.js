#!/usr/bin/env node
// @ts-check

const {Cli, Builtins} = require('clipanion');
const {BuildCommand, CleanCommand, DevCommand, DoctorCommand, InstallCommand} = require('../lib/cjs/cli/index.js');
const { fs } = require('@roots/bud-support')

const args = process.argv.splice(2);

(async () => {
  const manifest = await fs.readJson(`${__dirname}/../package.json`)
  const name = manifest.name.split('/').pop()

  const cli = new Cli({
    binaryLabel: name,
    binaryName: name,
    binaryVersion:manifest.version,
    enableColors: true,
  })

  cli.register(Builtins.HelpCommand)
  cli.register(Builtins.DefinitionsCommand)
  cli.register(Builtins.VersionCommand)
  cli.register(BuildCommand);
  cli.register(CleanCommand);
  cli.register(DevCommand);
  cli.register(DoctorCommand);
  cli.register(InstallCommand)

  cli.run(args);
})();
