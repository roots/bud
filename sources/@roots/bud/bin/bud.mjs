#!/usr/bin/env node

const run = async () => {
  await import(`../lib/cli/index.js`);
};

run();
