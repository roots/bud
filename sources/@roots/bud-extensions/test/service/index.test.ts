import {Bud, factory} from '@repo/test-kit/bud'
import type {Modules} from '@roots/bud-framework'
import type {ApplyPlugin} from '@roots/bud-framework/extension'
import {beforeEach, describe, expect, it, vi} from 'vitest'

import Extensions from '../../src/service/index.js'

describe(`@roots/bud-extensions`, () => {
  let bud: Bud
  let extensions: Extensions

  beforeEach(async () => {
    bud = await factory()
    extensions = new Extensions(() => bud)
  })

  it(`is constructable`, async () => {
    expect(extensions).toBeInstanceOf(Extensions)
  })

  it(`add fn registers a module`, async () => {
    extensions.repository = {} as any

    const options = {test: `foo`}

    const mockWebpackPlugin: ApplyPlugin = {apply: vi.fn()}

    const mockModule: any = {
      label: `mock_extension`,
      register: vi.fn(async () => null),
      boot: vi.fn(async () => null),
      options: options,
      make: vi.fn(async () => mockWebpackPlugin),
      when: vi.fn(async () => true),
    }

    await extensions.add(mockModule)

    const instance = extensions.get(`mock_extension` as keyof Modules)
    if (!instance) throw new Error(`Extension not found`)

    expect(instance.label).toBe(`mock_extension`)

    expect(extensions.get(mockModule.label)?.options?.test).toEqual(
      mockModule.options.test,
    )
  })

  it(`should assign a uuid as key for extensions without names`, async () => {
    extensions.repository = {} as any

    const mockExtension = {
      register: async () => {
        /*noop*/
      },
    }

    await extensions.add(mockExtension)

    expect(Object.keys(extensions.repository).sort().pop()).toMatch(
      /[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}/,
    )
  })

  it(`should accept a plugin signifier`, async () => {
    extensions.repository = {} as any

    // @ts-ignore
    await extensions.add(`palette-webpack-plugin`)

    expect(
      Object.values(extensions.repository).sort().pop().constructor.name,
    ).toBe(`PaletteWebpackPlugin`)
  })

  it(`should accept a plugin definition`, async () => {
    extensions.repository = {} as any

    const plugin = await import(`palette-webpack-plugin`)
    await extensions.add(plugin.default)

    expect(
      Object.values(extensions.repository).sort().pop().constructor.name,
    ).toBe(`PaletteWebpackPlugin`)
  })

  it(`should accept a plugin instance`, async () => {
    extensions.repository = {} as any

    const plugin = await import(`palette-webpack-plugin`)
    // @ts-ignore
    const instance = new plugin.default()
    await extensions.add(instance)

    expect(Object.values(extensions.repository).sort().pop()).toBe(
      instance,
    )
  })

  it(`should assign a uuid to label for extensions without name`, async () => {
    extensions.repository = {} as Extensions[`repository`]

    await extensions.add(
      // @ts-ignore
      {
        register: async () => {
          // noop
        },
      },
    )

    expect(
      // @ts-ignore
      Object.keys(extensions.repository).sort().pop(),
    ).toMatch(
      /[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}/,
    )
  })

  it(`bud.extensions.repository options should match snapshot in development`, async () => {
    bud = await factory({mode: `development`})

    const extensions = new Extensions(() => bud)
    if (!extensions) throw new Error(`Extensions not found`)

    if (!extensions.register) throw new Error(`Extensions.register not found`)
    await extensions.register(bud)

    if (!extensions.booted) throw new Error(`Extensions.booted not found`)
    await extensions.booted(bud)

    expect(Object.keys(extensions.repository).sort()).toMatchSnapshot()
  })
})
