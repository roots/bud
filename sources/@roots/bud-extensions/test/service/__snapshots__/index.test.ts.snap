// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@roots/bud-extensions > bud.extensions.repository options should match snapshot in development 1`] = `
[
  "@roots/bud-entrypoints",
  "@roots/bud-extensions/cdn",
  "@roots/bud-extensions/clean-webpack-plugin",
  "@roots/bud-extensions/copy-webpack-plugin",
  "@roots/bud-extensions/esm",
  "@roots/bud-extensions/fix-style-only-entrypoints",
  "@roots/bud-extensions/html-webpack-plugin",
  "@roots/bud-extensions/import-map",
  "@roots/bud-extensions/interpolate-html-webpack-plugin",
  "@roots/bud-extensions/mini-css-extract-plugin",
  "@roots/bud-extensions/tsconfig-values",
  "@roots/bud-extensions/webpack-define-plugin",
  "@roots/bud-extensions/webpack-hot-module-replacement-plugin",
  "@roots/bud-extensions/webpack-lifecycle-plugin",
  "@roots/bud-extensions/webpack-manifest-plugin",
  "@roots/bud-extensions/webpack-provide-plugin",
  "@roots/bud-minify",
  "@roots/bud-minify/minify-css",
  "@roots/bud-minify/minify-js",
  "@roots/bud-postcss",
  "@roots/bud-react",
  "@roots/bud-swc",
  "@roots/bud-tailwindcss",
]
`;
